// @cloudmpower/exceptionmanager_async/index.js
(function(){

    let utils = require("@cloudmpower/utility_async");

    let unExpectedException = function(config, tnxId){
        this.config = config;
        this.tnxId = tnxId;
        this.utils = utils;
        this.started = false;
    };

    unExpectedException.prototype = {
        enableExceptionHandling: function(){
            let resHandle = this.writeUnCaughtException.bind(this);
            process.on("uncaughtException", resHandle);
            process.on("error", resHandle);
            this.appStarted();
        },
        enableAppExceptionHandling: function(app){
            let resHandle = this.writeAppException.bind(this);
            app.on("error", resHandle);
            this.appStarted();
        },
        writeUnCaughtException: function(err){
            let error = err.stack;
            this.utils.log(error , "process unExpectedError", this.config);
        },
        writeAppException: function(err , ctx){
            let error = err.stack;
            this.utils.log(error, "appException", this.config);

            let responseError = { "status": 409 , "responseData": { "errorCode": "ERR409", "message": "Incorrect Request"}};
            if(ctx.method == "POST") {
                ctx.status = 409;
                ctx.res.writeHead(409, {"Content-Type": "application/json"});
                ctx.res.write(JSON.stringify(responseError));
                ctx.res.end();
            }
        },
        appStarted: function(){
            if(!this.started) this.utils.log("listening to " + this.config.appPort, "AppStarted", this.config);
            this.started = true;
        }
    };

    module.exports = function(config, tnxId){
        return (new unExpectedException(config, tnxId));
    };
})();
