// @cloudmpower/interactor_async/index.js
(function(){

    var fs = require("fs");

    var path = require("path");

    var _ = require("lodash");

    var responseHandler = require("@cloudmpower/responsehandler_async");

    var utils = require("@cloudmpower/utility_async");

    var dbAbstrator = require("@cloudmpower/dbabstraction_async");

    var serverBlocker = require("@cloudmpower/serverblocker_async");

    var serverConfig = require("@cloudmpower/serverconfig_async");

    var consoleUpgrade = require("@cloudmpower/console_upgrader_async");

    var referenceObj = [];

    var fileNames = {
        _                           : "_",
        fs                          : "fs",
        path                        : "path",
        apiLoader                   : "apiLoader",
        responseHandler             : "responseHandler",
        utils                       : "utils",
        dbAbstrator                 : "dbAbstrator",
        serverBlocker               : "serverBlocker",
        serverConfig                : "serverConfig",
        consoleUpgrade              : "consoleUpgrade"
    };

    var loadOtherModules = function(){

        referenceObj = [
            {
                fileName        : "_",
                fileObj         : _
            },
            {
                fileName        : "fs",
                fileObj         : fs
            },
            {
                fileName        : "path",
                fileObj         : path
            },
            {
                fileName        : "utils",
                fileObj         : utils
            },
            {
                fileName        : "responseHandler",
                fileObj         : responseHandler
            },
            {
                fileName        : "dbAbstrator",
                fileObj         : dbAbstrator
            },
            {
                fileName        : "serverBlocker",
                fileObj         : serverBlocker
            },
            {
                fileName        : "serverConfig",
                fileObj         : serverConfig
            },
            {
                fileName        : "consoleUpgrade",
                fileObj         : consoleUpgrade
            }
        ];
    };

    var getRequiredFile = function(){};

    getRequiredFile.prototype = {

        returnCorrectFile: function(fileName){
            var resHandle = this.incorrectFile.bind(this);
            var refObj = _.find(referenceObj, {fileName: fileName});
            return (typeof refObj == "undefined")? resHandle : refObj.fileObj;
        },
        incorrectFile: function(){
            console.log("Incorrect File required");
            return false;
        },
        requireExtension: function(extendFileObj, done){
            extendFileObj = (typeof extendFileObj != "object") ? []: extendFileObj;
            referenceObj = _.union(extendFileObj, referenceObj);

            for(var i = 0; i < extendFileObj.length; i++){
                fileNames[extendFileObj[i].fileName] = extendFileObj[i].fileName;
            }
            if(done) done();
        },
        getFileName: function(){
            return fileNames;
        }
    };

    module.exports = function(fileName){
        var file = new getRequiredFile();
        return (file.returnCorrectFile(fileName));
    };

    module.exports.requireHandler = function(extended){
        var file = new getRequiredFile();
        return (file.requireExtension(extended));
    };

    module.exports.fileName = function(){
        var file = new getRequiredFile();
        return (file.getFileName());
    };

    loadOtherModules();
})();
